FaaSController (#59)
megaease/easegress: doc/faascontroller.md

# FaaSController 
- [FaaSController](#faascontroller)
  - [Prerequest](#prerequest)
  - [Configuration](#configuration)
    - [Controller spec](#controller-spec)
    - [FaaSFunction spec](#faasfunction-spec)
    - [Lifecycle](#lifecycle)
    - [RESTful APIs](#restful-apis)
  - [Demoing](#demoing)
  - [Reference](#reference)

* A FaaSController is a business controller for handling Easegress and FaaS products integration purposes.  It abstracts `FaasFunction`, `FaaSStore` and, `FaaSProvder`. Currently, we only support `Knative` type FaaSProvider. The `FaaSFunction` describes the name, image URL, the resource, and autoscaling type of this FaaS function instance. The `FaaSStore` is covered by Easegress' embed ETCD already.

> * FaaSController works closely with local FaaSProvider. Please make sure they are running in a communicable environment. Flow this [doc](https://knative.dev/docs/install/install-serving-with-yaml/) to install Knative[1]'s serving component in K8s. It's better to have Easegress run in the same vm-instances with K8s for saving communication costs.

Author: Fix, updated.

> * FaaSController works closely with local FaaSProvider. Please make sure the they are running int the communicable environment. Flow this [doc](https://knative.dev/docs/install/install-serving-with-yaml/) to install Knative[1]'s serving component in K8s. It's better to have Easegress run in the same vm-instances with K8s for saving communication cost.

Reviewer: ```suggestion
* FaaSController works closely with local FaaSProvider. Please make sure they are running in a communicable environment. Flow this [doc](https://knative.dev/docs/install/install-serving-with-yaml/) to install Knative[1]'s serving component in K8s. It's better to have Easegress run in the same vm-instances with K8s for saving communication costs.
```

Reviewer: propose to use a grammar tool, as the one @xxx7xxxx proposed to me: https://app.grammarly.com/

* After updating the function, it will run into `pending` status. (Once provision successfully, it will become `active` automatically) 
* FaaSController will remove the routing rule in its HTTPServer for rejecting one function's ingress traffic when it's in `inactive` status. (Then the client will receive HTTP 404 failure response. For becoming zero-downtime, deploy another new FaaSFunction in this FaaSController may be helpful)

```                                                      
                                                         +-------------------+
                                                         | provision succ    |
  +---------------+   provision succ              +------+------+<-----------+
  |               |------------------------------>|             |
  |    pending    |             provision pending |   active    |<------+
  |               | <-----------------------------|             |       |
+-+---------------+ <--+     +--------------------+--------+----+       |   
|                      |     |                             |            |

> provision faild        |     |                             |stop        | start

Reviewer 2: failed

>                                                          +-------------------+
>                                                          | provision succ    |
>   +---------------+   provision succ              +------+------+<-----------+
>   |               |------------------------------>|             |
>   |    pending    |             provision pending |   active    |<------+
>   |               | <-----------------------------|             |       |
> +-+---------------+ <--+     +--------------------+--------+----+       |   
> |                      |     |                             |            |
> provision faild        |     |                             |stop        | start
> |                  update    provision failed              |            |
> |  +---------------+   |     |        +-------------+      |            |
> +->|               |   |     |        |             |      |            |
>    |   failed      |<--+-----+--------+   inactive  |<-----+            |
>    |               |                  |             +-------------------+
>    |               |                  |             |
>    +---------------+                  +-------------+

Reviewer 3: ```suggestion
                  Provision

                      │
                      │
                      │
                ┌─────▼──────┐      Start     ┌────────────┐
                │            │     Success    │            │
       ┌────────┤  Pending   ├────────────────►   Active   │
       │        │            │                │            ├──────┐
       │        └───┬───▲────┘                └────┬───▲───┘      │
       │            │   │                          │   │          │
       │            │   │                          │   │          │
       │      Start │   │ Update             Stop  │   │ Start    │
       │     Failed │   │                          │   │Success   │
       │            │   │                          │   │          │
       │        ┌───▼───┴────┐                ┌────▼───┴───┐      │
       │        │            │                │            │      │
Delete │        │   Failed   ◄────────────────┤  Inactive  │      │
       │        │            │  Start Failed  │            │      │
       │        └───┬───▲────┘                └──────┬─────┘      │
       │            │   │                            │            │
       │            │   │              Errors        │            │
       │    Delete  │   └────────────────────────────┼────────────┘
       │            │                                │
       │            │                                │
       │        ┌───▼────────┐                       │
       │        │            │          Delete       │
       └────────►  Destory   ◄───────────────────────┘
                │            │
                └────────────┘
```

Reviewer 3: The original state machine is not right. I proposed this one.

Author: All related suggestions have been updated manually. Thx for correcting.

* `Initial`: Once the function has been created in Easegress, its original state is `initial`. After checking FaaSProvider(Knative)'s status successfully, it will become `active` automatically. And the function is ready for handling traffic.

* `Active`: Easegress's FaaSFunction will be `active` not matter there are requests or not.   **Easegress will only route ingress traffic to FaaSProvider when the function is in the `active` state.**

* `Inactive`: Stopping function execution by calling FaaSController's `stop` RESTful API and it will run into `inactive`. Updating function's spec for image URL or other fields, or deleting function also need to stop it first.

* `Failed`: The function will be turned into `failed` states during the runtime checking. If it's about some configuration error, e.g., wrong docker image URL, we can detect this failure by function's `status` message and then update the function's spec by calling RESTful API. If it's about some temporal failure caused by FaaSProvider, the function will turn into the `initial` state after FaaSProvider is recovered.

```                                                      

>                                   +---------------+     start succ                  +---------------+
>                                   |               |-------------------------------->|               |
> Provision ----------------------->|    initial    |                                 |    active     |<------------+
>                    +--------------+               |                                 |               |             |
>                    |            +-+---------------+ <----+          +---------------+---------------+             |   
>                    |            |                        |          |                               |             |
>                    |            |              +---------+        errors                            |             |
>                    |         errors         update       |          |                              stop         start succ 
>                    |            |        (check succ)  update       |                               |             |
>                    |            |              |         |          |                               |             |  
>                    |            |  +-----------+---+     +----------+-----+---------------+         |             |
>                    |            +->|               |                |     |               |         |             |
>                 delete             |    failed     |<-----------errors----+   inactive    |<--------+             |
>                    |               |               |                      |               +-----------------------+
>                    |               +---------+-----+                      +-------+-------+
>                    |                         |                                    | 
>                    |    +---------------+  delete                               delete 
>                    +--->|               |    |                                    |
>                         |   Destroyed   +<---+------------------------------------+
>                         |               |
>                         +---------------+

Reviewer 3: ```suggestion
                  Provision

                      │
                      │
                      │
                ┌─────▼──────┐      Start     ┌────────────┐
                │            │     Success    │            │
       ┌────────┤  Initial   ├────────────────►   Active   │
       │        │            │                │            ├──────┐
       │        └───┬───▲────┘                └────┬───▲───┘      │
       │            │   │                          │   │          │
       │            │   │                          │   │          │
       │      Errors│   │ Update             Stop  │   │ Start    │
       │            │   ├───────────┐              │   │Success   │
       │            │   │           │              │   │          │
       │        ┌───▼───┴────┐      │         ┌────▼───┴───┐      │
       │        │            │      └─────────┤            │      │
Delete │        │   Failed   │                │  Inactive  │      │
       │        │            ◄────────────────┤            │      │
       │        └───┬───▲────┘  Start Failed  └──────┬─────┘      │
       │            │   │                            │            │
       │            │   │              Errors        │            │
       │    Delete  │   └────────────────────────────┼────────────┘
       │            │                                │
       │            │                                │
       │        ┌───▼────────┐                       │
       │        │            │          Delete       │
       └────────►  Destory   ◄───────────────────────┘
                │            │
                └────────────┘
```
